cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultImpersonator.java

```diff
@@ -83,14 +86,15 @@ private UserGroupInformation getUGI(NamespacedEntityId entityId,
     }
 
     ImpersonationRequest impersonationRequest = new ImpersonationRequest(entityId, impersonatedOpType);
-    UserGroupInformation configuredUGI = ugiProvider.getConfiguredUGI(impersonationRequest).getUGI();
-    // Only cdap master user is allowed to impersonate another user besides itself. See CDAP-8641
-    String currUserShortName = UserGroupInformation.getCurrentUser().getShortUserName();
-    if (currUserShortName.equals(masterShortUsername) || currUserShortName.equals(configuredUGI.getShortUserName())) {
-      return configuredUGI;
+    // if the current user is not same as cdap master user then it means we are already impersonating some user
+    // and hence we should not allow another impersonation. See CDAP-8641
+    if (!UserGroupInformation.getCurrentUser().getShortUserName().equals(masterShortUsername)) {
```

Comment:
Would be good to store `UserGroupInformation.getCurrentUser()` in a temp variable so that we don't have to make multiple calls.